### PR TITLE
Molecule Driver

### LINKED ISSUE #165: Extract `MoleculeDriver`
### What is the expected enhancement?

As part of #147 I propose to extract the common `Molecule`-based driver initialization into a separate driver implementation. As suggested in the initial proposal, this could look similar to the following:
```python
class MoleculeDriver:
   def __init__(self, molecule, driver_type, driver_kwargs):
      # intitialize stuff

   def run(self):
      driver_cls = getattr(qiskit_nature.drivers, self._driver_type)
      driver = driver_cls.from_molecule(self._molecule)
      for key, value in self._driver_kwargs:
         setattr(driver, key, value)
      return driver.run()
```
More details will need to be figured out as we proceed with the actual implementation in the upcoming weeks.

### PR DESCRIPTION
**Pull request by manoelmarques**
### Summary
Fixes #165

### Details and comments
The native drivers were refactored to separate the molecule handling code and have a class factory method `from_molecule`.

Classes `FermionicMoleculeDriver` and `BosonicMoleculeDriver` created.

The drivers were updated to use MissingOptionalLibraryError when not installed. The existent unit tests were simplified with the decorator `requires_extra_library`.

### CODE REVIEW

File: docs/tutorials/03_ground_state_solvers.ipynb
Comment (mrossinek) [inline docs/tutorials/03_ground_state_solvers.ipynb:42]: Since this h5py deprecation from within PySCF is a known issue (#248, https://github.com/pyscf/pyscf/issues/1003) I am wondering whether we should catch it (or at least locally in this tutorial)? :thinking:

File: docs/tutorials/03_ground_state_solvers.ipynb
Comment (mrossinek) [inline docs/tutorials/03_ground_state_solvers.ipynb:268]: I am curious why this value changed so significantly.. I don't think it's this PRs fault but maybe we should look into that!

File: qiskit_nature/drivers/second_quantization/bosonic_molecule_driver.py
Comment (mrossinek) [inline qiskit_nature/drivers/second_quantization/bosonic_molecule_driver.py:37]: Missing `Args` and `Raises` documentation
Missing returned mypy type

File: qiskit_nature/drivers/second_quantization/fermionic_driver.py
Comment (mrossinek) [inline qiskit_nature/drivers/second_quantization/fermionic_driver.py:46]: Why do you declare this as abstract?

File: qiskit_nature/drivers/second_quantization/gaussiand/gaussian_forces_driver.py
Comment (mrossinek) [inline qiskit_nature/drivers/second_quantization/gaussiand/gaussian_forces_driver.py:49]: Can we remove these args in all drivers without proper deprecation?

File: qiskit_nature/drivers/second_quantization/gaussiand/gaussian_forces_driver.py
Comment (mrossinek) [inline qiskit_nature/drivers/second_quantization/gaussiand/gaussian_forces_driver.py:114]: Why not move the code from `check_valid()` into here?

File: qiskit_nature/drivers/second_quantization/base_driver.py
Comment (mrossinek) [inline qiskit_nature/drivers/second_quantization/base_driver.py:67]: I don't really understand the point of this line..
Do we even still need `supports_molecule` or can we simply rely on whether `from_molecule` is implemented?

**Review by mrossinek — COMMENTED** [review]
Thanks @manoelmarques! I have some (mostly minor) comments :slightly_smiling_face: Otherwise this LGTM :+1:

File: qiskit_nature/drivers/second_quantization/gaussiand/gaussian_forces_driver.py
Comment (manoelmarques) [reply qiskit_nature/drivers/second_quantization/gaussiand/gaussian_forces_driver.py:49]: Those drivers haven't been released yet. There is no need to deprecate unreleased code.

File: qiskit_nature/drivers/second_quantization/gaussiand/gaussian_forces_driver.py
Comment (manoelmarques) [reply qiskit_nature/drivers/second_quantization/gaussiand/gaussian_forces_driver.py:114]: check_valid() is part of gaussian_utils and use local variables that are used also by another function there (run_g16).
Moving it to the two Gaussian drivers would mean copy all this code to each driver.

File: qiskit_nature/drivers/second_quantization/gaussiand/gaussian_forces_driver.py
Comment (mrossinek) [reply qiskit_nature/drivers/second_quantization/gaussiand/gaussian_forces_driver.py:49]: Ah because we relocated them, right. But then we should adapt the deprecation messages in the old locations. Otherwise they suggest the drivers can be used as direct replacements (which now will no longer be the case, depending on the use case)

File: qiskit_nature/drivers/second_quantization/gaussiand/gaussian_forces_driver.py
Comment (mrossinek) [reply qiskit_nature/drivers/second_quantization/gaussiand/gaussian_forces_driver.py:114]: Ah I hadn't noticed that. Thanks!

File: docs/tutorials/03_ground_state_solvers.ipynb
Comment (manoelmarques) [reply docs/tutorials/03_ground_state_solvers.ipynb:42]: In the issue above: "This has been fixed in Pyscf v2.0.0.a, although it's not on Pypi yet."
So they already addressed it and it should be in their next release.

File: qiskit_nature/drivers/second_quantization/bosonic_molecule_driver.py
Comment (manoelmarques) [reply qiskit_nature/drivers/second_quantization/bosonic_molecule_driver.py]: Added the missing docs and type

File: qiskit_nature/drivers/second_quantization/fermionic_driver.py
Comment (manoelmarques) [reply qiskit_nature/drivers/second_quantization/fermionic_driver.py]: Removed the abstract

File: qiskit_nature/drivers/second_quantization/gaussiand/gaussian_forces_driver.py
Comment (manoelmarques) [reply qiskit_nature/drivers/second_quantization/gaussiand/gaussian_forces_driver.py:49]: Fixed deprecation msgs for the drivers that changed

File: qiskit_nature/drivers/second_quantization/base_driver.py
Comment (manoelmarques) [reply qiskit_nature/drivers/second_quantization/base_driver.py]: Fixed supports_molecule as the above was bogus. Since other methods in Nature use argument of type BaseDriver and access the molecule property, it is still needed so that they fail when try to access the property for a driver that doesn't support it.

File: docs/tutorials/03_ground_state_solvers.ipynb
Comment (manoelmarques) [reply docs/tutorials/03_ground_state_solvers.ipynb:268]: I ran locally from MacOS but those are similar to results we have been getting from 'main'. Looking at the tutorials run from a recent nightly CI (Ubuntu):
```
Vibrational ground state energy (cm^-1): (2555.376153240098-0j)
```

File: docs/tutorials/03_ground_state_solvers.ipynb
Comment (mrossinek) [reply docs/tutorials/03_ground_state_solvers.ipynb:42]: I was just thinking that the tutorials could look cleaner in the meantime. v2.0.0a has been around for a while and I don't see a timeline for when 2.0.0 will be published on PyPI..

But if you think it's fine as is, then it's okay :+1:

File: docs/tutorials/03_ground_state_solvers.ipynb
Comment (mrossinek) [reply docs/tutorials/03_ground_state_solvers.ipynb:268]: Yeh this could have happened during the migration from Aqua. But I think nonetheless we should investigate when/why this differences started to occur. Maybe a bisection will tell :thinking: (But this is unrelated to this PR for sure!)

File: qiskit_nature/drivers/second_quantization/fermionic_molecule_driver.py
Comment (mrossinek) [inline qiskit_nature/drivers/second_quantization/fermionic_molecule_driver.py:160]: I am wondering whether it would make sense to define a small `MoleculeDriver` interface which would take care of all these properties (which are identical in here and in `BosonicMoleculeDriver`). This would also allow type checking against the interface in case we ever need to differentiate a `MoleculeDriver` from another one (say e.g. in the `BopesSampler`)

File: qiskit_nature/drivers/second_quantization/gaussiand/gaussiandriver.py
Comment (mrossinek) [inline qiskit_nature/drivers/second_quantization/gaussiand/gaussiandriver.py:110]: Why doesn't this return a `str`?

File: qiskit_nature/drivers/second_quantization/pyquanted/pyquantedriver.py
Comment (mrossinek) [inline qiskit_nature/drivers/second_quantization/pyquanted/pyquantedriver.py:43]: What's the return type here?

**Review by mrossinek — COMMENTED** [review]
Some minor comments but otherwise this is looking good :+1:

File: qiskit_nature/drivers/second_quantization/fermionic_molecule_driver.py
Comment (manoelmarques) [reply qiskit_nature/drivers/second_quantization/fermionic_molecule_driver.py:160]: I had discussed that with Steve but it would need to be a separate interface not derived from BaseDriver. so it would be 2 base classes for those and you wouldn't be able to use as a typehint anyway because it is not a BaseDriver

File: qiskit_nature/drivers/second_quantization/gaussiand/gaussiandriver.py
Comment (manoelmarques) [reply qiskit_nature/drivers/second_quantization/gaussiand/gaussiandriver.py]: Added docstring and typehints

File: qiskit_nature/drivers/second_quantization/pyquanted/pyquantedriver.py
Comment (manoelmarques) [reply qiskit_nature/drivers/second_quantization/pyquanted/pyquantedriver.py]: Added docstring and typehints

File: qiskit_nature/drivers/second_quantization/fermionic_molecule_driver.py
Comment (mrossinek) [reply qiskit_nature/drivers/second_quantization/fermionic_molecule_driver.py:160]: What is the problem with using two base classes?

File: qiskit_nature/drivers/second_quantization/electronic_structure_driver.py
Comment (mrossinek) [inline qiskit_nature/drivers/second_quantization/electronic_structure_driver.py:40]: I think the deprecation of `qiskit_nature.drivers.fermionic_driver.FermionicDriver` is still missing. Same for the `BosonicDriver` on that level.

File: qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py
Comment (mrossinek) [inline qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py:39]: I am not sure whether that will work but rather than having the values be strings, could we not set the classes here directly?
Below you could then do something along the lines of:
```python
if driver_type == ElectronicStructureDriverType.AUTO:
    ...
else:
    driver_class = driver_type.value
    driver_class.check_installed()
```

File: qiskit_nature/drivers/second_quantization/fermionic_molecule_driver.py
Comment (mrossinek) [reply qiskit_nature/drivers/second_quantization/fermionic_molecule_driver.py]: The idea was also whether we could extract the duplication of the attributes like `basis`, `method`, etc.
But I guess that is fine as is..

File: qiskit_nature/drivers/second_quantization/electronic_structure_driver.py
Comment (manoelmarques) [reply qiskit_nature/drivers/second_quantization/electronic_structure_driver.py:40]: I added the deprecations

File: qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py
Comment (mrossinek) [reply qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py]: Thanks! I see now that I was simplifying it a bit too much because my proposal would have involved importing all drivers :sweat_smile: But the current approach looks good to me as it simplifies future addition of drivers :+1:

**Review by mrossinek — DISMISSED** [review]
I think this looks good to me now! Thanks @manoelmarques :+1: :tada: 

Before merging let's await a second review though

File: qiskit_nature/drivers/second_quantization/gaussiand/gaussiandriver.py
Comment (woodsp-ibm) [inline qiskit_nature/drivers/second_quantization/gaussiand/gaussiandriver.py:123]: Since check_valid is `-> None` should the return type here be that too otherwise this almost feels like a boolean. Maybe it should list what it does as a failure with a Raises here? (Same comment I think applies to other places that are similar)

File: qiskit_nature/drivers/second_quantization/pyscfd/pyscfdriver.py
Comment (woodsp-ibm) [inline qiskit_nature/drivers/second_quantization/pyscfd/pyscfdriver.py:215]: For PySCF it now supports DFT methods too. We can shorten it to say `Returns method` or have it be a be more than stating what the code already tells you by saying `Returns Hartree-Fock/Kohn-Sham method`  @mrossinek any thought

**Comment by woodsp-ibm** [comment]
Aside from those minor comments it looks good to me.

File: qiskit_nature/drivers/second_quantization/pyscfd/pyscfdriver.py
Comment (manoelmarques) [reply qiskit_nature/drivers/second_quantization/pyscfd/pyscfdriver.py:215]: I added the typehint return and completed the docstrings with the exception

File: qiskit_nature/drivers/second_quantization/gaussiand/gaussiandriver.py
Comment (manoelmarques) [reply qiskit_nature/drivers/second_quantization/gaussiand/gaussiandriver.py]: I added the typehint return and completed the docstring with the exception

File: qiskit_nature/drivers/second_quantization/pyscfd/pyscfdriver.py
Comment (mrossinek) [reply qiskit_nature/drivers/second_quantization/pyscfd/pyscfdriver.py:215]: Yes that is true. I think the suggested docstring should be fine :+1:

Comment (manoelmarques) [reply qiskit_nature/drivers/second_quantization/pyscfd/pyscfdriver.py]: Changed the docstring

**Review by mrossinek — APPROVED** [review]
This LGTM! Thanks, @manoelmarques! :slightly_smiling_face: 

Edit: well I should have checked the CI first.. :see_no_evil: I will re-approve once the spelling is fixed :+1:

**Review by mrossinek — APPROVED** [review]

**Pull request merged**
